pgmpy/base/base.py

```diff
@@ -0,0 +1,248 @@
+import inspect
+from abc import ABC, abstractmethod
+from typing import Hashable, Iterable, Optional
+
+import networkx as nx
+
+from pgmpy.base._mixin_roles import _GraphRolesMixin
+
+
+class _CoreGraphABC(ABC):
+    """
+    An abstract class.
+
+    Inheriting from `CoreGraph` automatically includes inheritance from `_CoreGraphABC`.
+    """
+
+    # Constants that the graph class must implement.
+    ALLOWED_EDGES = None
+    IS_DIRECTED = None
+    IS_MULTIGRAPH = None
+
+    def __init_subclass__(cls, **kwargs):
+        super().__init_subclass__(**kwargs)
+
+        if inspect.isabstract(cls):
+            return
+
+        if cls.IS_DIRECTED is None:
+            raise TypeError("IS_DIRECTED must be defined.")
+
+        if cls.IS_MULTIGRAPH is None:
+            raise TypeError("IS_MULTIGRAPH must be defined.")
+
+        if cls.ALLOWED_EDGES is None:
+            raise TypeError("ALLOWED_EDGES must be defined.")
+
+    @abstractmethod
+    def _graph_type_marker(self):
+        """
+        This is a pure virtual method to prevent this class (and CoreGraph) from being instantiated directly.
+        Concrete classes like DAG, PDAG, etc., must implement it.
+        """
+        pass
+
+
+class CoreGraph(_CoreGraphABC, nx.MultiGraph, _GraphRolesMixin):
```

Comment:
Would suggest renaming this to `_CoreGraph` to signify that it is an internal/private class.